File: weather.go
// weather.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"time"
)

var apiKey = os.Getenv("OPENWEATHER_API_KEY")
var units = "metric"
var cache = make(map[string]cacheEntry)
const cacheTTL = 600

type cacheEntry struct {
	data      interface{}
	timestamp time.Time
}

type weatherData struct {
	Weather []struct {
		Main        string `json:"main"`
		Description string `json:"description"`
		Icon        string `json:"icon"`
	} `json:"weather"`
	Main struct {
		Temp      float64 `json:"temp"`
		FeelsLike float64 `json:"feels_like"`
		Humidity  int     `json:"humidity"`
		Pressure  int     `json:"pressure"`
	} `json:"main"`
	Wind struct {
		Speed float64 `json:"speed"`
	} `json:"wind"`
	Name string `json:"name"`
	Sys  struct {
		Country string `json:"country"`
	} `json:"sys"`
	Dt int64 `json:"dt"`
	Cod int   `json:"cod"`
}

type forecastData struct {
	Cod string `json:"cod"`
	List []struct {
		Dt   int64 `json:"dt"`
		Main struct {
			TempMin float64 `json:"temp_min"`
			TempMax float64 `json:"temp_max"`
		} `json:"main"`
		Weather []struct {
			Icon string `json:"icon"`
		} `json:"weather"`
	} `json:"list"`
}

func getWeather(city string) (*weatherData, error) {
	cacheKey := "weather_" + city + "_" + units
	if entry, ok := cache[cacheKey]; ok {
		if time.Since(entry.timestamp).Seconds() < cacheTTL {
			return entry.data.(*weatherData), nil
		}
	}
	base := "https://api.openweathermap.org/data/2.5/weather"
	params := url.Values{}
	params.Set("q", city)
	params.Set("appid", apiKey)
	params.Set("units", units)
	fullURL := base + "?" + params.Encode()
	resp, err := http.Get(fullURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var w weatherData
	if err := json.Unmarshal(body, &w); err != nil {
		return nil, err
	}
	cache[cacheKey] = cacheEntry{data: &w, timestamp: time.Now()}
	return &w, nil
}

func getForecast(city string) (*forecastData, error) {
	cacheKey := "forecast_" + city + "_" + units
	if entry, ok := cache[cacheKey]; ok {
		if time.Since(entry.timestamp).Seconds() < cacheTTL {
			return entry.data.(*forecastData), nil
		}
	}
	base := "https://api.openweathermap.org/data/2.5/forecast"
	params := url.Values{}
	params.Set("q", city)
	params.Set("appid", apiKey)
	params.Set("units", units)
	fullURL := base + "?" + params.Encode()
	resp, err := http.Get(fullURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var f forecastData
	if err := json.Unmarshal(body, &f); err != nil {
		return nil, err
	}
	cache[cacheKey] = cacheEntry{data: &f, timestamp: time.Now()}
	return &f, nil
}

func iconForCode(code string) string {
	m := map[string]string{
		"01d": "☀️", "01n": "🌙",
		"02d": "⛅", "02n": "☁️",
		"03d": "☁️", "03n": "☁️",
		"04d": "☁️", "04n": "☁️",
		"09d": "🌧️", "09n": "🌧️",
		"10d": "🌦️", "10n": "🌧️",
		"11d": "⛈️", "11n": "⛈️",
		"13d": "❄️", "13n": "❄️",
		"50d": "🌫️", "50n": "🌫️",
	}
	if val, ok := m[code]; ok {
		return val
	}
	return "🌈"
}

func displayWeather(w *weatherData) {
	if w == nil || w.Cod != 200 {
		fmt.Println("Could not retrieve weather.")
		return
	}
	icon := iconForCode(w.Weather[0].Icon)
	dt := time.Unix(w.Dt, 0).Format("2006-01-02 15:04")
	tempUnit := "°C"
	windUnit := "m/s"
	if units == "imperial" {
		tempUnit = "°F"
		windUnit = "mph"
	}
	fmt.Printf("\n%s Weather in %s%s (%s)\n", icon, w.Name, func() string {
		if w.Sys.Country != "" {
			return ", " + w.Sys.Country
		}
		return ""
	}(), dt)
	fmt.Printf("Temperature: %.1f%s (feels like %.1f%s)\n", w.Main.Temp, tempUnit, w.Main.FeelsLike, tempUnit)
	fmt.Printf("Humidity: %d%%   Pressure: %d hPa\n", w.Main.Humidity, w.Main.Pressure)
	fmt.Printf("Wind: %.1f %s\n", w.Wind.Speed, windUnit)
	fmt.Printf("Description: %s\n", w.Weather[0].Description)
}

func displayForecast(f *forecastData) {
	if f == nil || f.Cod != "200" {
		return
	}
	fmt.Println("\n📅 5‑Day Forecast:")
	days := make(map[string]struct{ Min, Max float64; Icon string })
	for _, entry := range f.List[:40] {
		dt := time.Unix(entry.Dt, 0)
		dateKey := dt.Format("2006-01-02")
		if _, ok := days[dateKey]; !ok {
			days[dateKey] = struct{ Min, Max float64; Icon string }{
				Min:  entry.Main.TempMin,
				Max:  entry.Main.TempMax,
				Icon: entry.Weather[0].Icon,
			}
		} else {
			d := days[dateKey]
			if entry.Main.TempMin < d.Min {
				d.Min = entry.Main.TempMin
			}
			if entry.Main.TempMax > d.Max {
				d.Max = entry.Main.TempMax
			}
			days[dateKey] = d
		}
	}
	tempUnit := "°C"
	if units == "imperial" {
		tempUnit = "°F"
	}
	count := 0
	for date, vals := range days {
		if count >= 5 {
			break
		}
		icon := iconForCode(vals.Icon)
		fmt.Printf("%s: %s %.0f%s / %.0f%s\n", date, icon, vals.Max, tempUnit, vals.Min, tempUnit)
		count++
	}
}

func main() {
	if apiKey == "" {
		fmt.Println("Please set OPENWEATHER_API_KEY environment variable.")
		os.Exit(1)
	}
	if os.Getenv("WEATHER_UNIT") == "imperial" {
		units = "imperial"
	}
	city := "London"
	if len(os.Args) > 1 {
		city = os.Args[1]
	}
	weather, err := getWeather(city)
	if err != nil {
		fmt.Println("Error getting weather:", err)
		return
	}
	displayWeather(weather)
	forecast, err := getForecast(city)
	if err == nil {
		displayForecast(forecast)
	}
}

File: weather.swift
// weather.swift
import Foundation

let apiKey = ProcessInfo.processInfo.environment["OPENWEATHER_API_KEY"] ?? "YOUR_API_KEY_HERE"
let units = ProcessInfo.processInfo.environment["WEATHER_UNIT"] == "imperial" ? "imperial" : "metric"
var cache: [String: (data: Any, timestamp: TimeInterval)] = [:]
let cacheTTL: TimeInterval = 600

func fetchData(urlString: String) -> Any? {
    guard let url = URL(string: urlString) else { return nil }
    let semaphore = DispatchSemaphore(value: 0)
    var result: Any? = nil
    let task = URLSession.shared.dataTask(with: url) { data, _, error in
        defer { semaphore.signal() }
        guard let data = data, error == nil else { return }
        do {
            result = try JSONSerialization.jsonObject(with: data, options: [])
        } catch {
            print("JSON error: \(error)")
        }
    }
    task.resume()
    semaphore.wait()
    return result
}

func getWeather(city: String) -> Any? {
    let key = "weather_\(city)_\(units)"
    if let entry = cache[key] {
        if Date().timeIntervalSince1970 - entry.timestamp < cacheTTL {
            return entry.data
        }
    }
    let url = "https://api.openweathermap.org/data/2.5/weather?q=\(city.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? city)&appid=\(apiKey)&units=\(units)"
    let data = fetchData(urlString: url)
    if let data = data {
        cache[key] = (data: data, timestamp: Date().timeIntervalSince1970)
    }
    return data
}

func getForecast(city: String) -> Any? {
    let key = "forecast_\(city)_\(units)"
    if let entry = cache[key] {
        if Date().timeIntervalSince1970 - entry.timestamp < cacheTTL {
            return entry.data
        }
    }
    let url = "https://api.openweathermap.org/data/2.5/forecast?q=\(city.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? city)&appid=\(apiKey)&units=\(units)"
    let data = fetchData(urlString: url)
    if let data = data {
        cache[key] = (data: data, timestamp: Date().timeIntervalSince1970)
    }
    return data
}

func iconForCode(_ code: String) -> String {
    let map: [String: String] = [
        "01d": "☀️", "01n": "🌙",
        "02d": "⛅", "02n": "☁️",
        "03d": "☁️", "03n": "☁️",
        "04d": "☁️", "04n": "☁️",
        "09d": "🌧️", "09n": "🌧️",
        "10d": "🌦️", "10n": "🌧️",
        "11d": "⛈️", "11n": "⛈️",
        "13d": "❄️", "13n": "❄️",
        "50d": "🌫️", "50n": "🌫️"
    ]
    return map[code] ?? "🌈"
}

func displayWeather(_ data: Any?) {
    guard let json = data as? [String: Any],
          let cod = json["cod"] as? Int, cod == 200 else {
        print("Could not retrieve weather.")
        return
    }
    let city = json["name"] as? String ?? "Unknown"
    let country = (json["sys"] as? [String: Any])?["country"] as? String ?? ""
    let dt = Date(timeIntervalSince1970: json["dt"] as? TimeInterval ?? 0)
    let dateFormatter = DateFormatter()
    dateFormatter.dateFormat = "yyyy-MM-dd HH:mm"
    let dtStr = dateFormatter.string(from: dt)
    let weatherArr = json["weather"] as? [[String: Any]] ?? []
    let weather = weatherArr.first ?? [:]
    let icon = iconForCode(weather["icon"] as? String ?? "")
    let main = json["main"] as? [String: Any] ?? [:]
    let temp = main["temp"] as? Double ?? 0
    let feels = main["feels_like"] as? Double ?? 0
    let humidity = main["humidity"] as? Int ?? 0
    let pressure = main["pressure"] as? Int ?? 0
    let wind = (json["wind"] as? [String: Any])?["speed"] as? Double ?? 0
    let desc = weather["description"] as? String ?? ""
    let tempUnit = units == "metric" ? "°C" : "°F"
    let windUnit = units == "metric" ? "m/s" : "mph"
    print("\n\(icon) Weather in \(city)\(country.isEmpty ? "" : ", " + country) (\(dtStr))")
    print("Temperature: \(String(format: "%.1f", temp))\(tempUnit) (feels like \(String(format: "%.1f", feels))\(tempUnit))")
    print("Humidity: \(humidity)%   Pressure: \(pressure) hPa")
    print("Wind: \(String(format: "%.1f", wind)) \(windUnit)")
    print("Description: \(desc.capitalized)")
}

func displayForecast(_ data: Any?) {
    guard let json = data as? [String: Any],
          let cod = json["cod"] as? String, cod == "200" else { return }
    let list = json["list"] as? [[String: Any]] ?? []
    var days: [String: (min: Double, max: Double, icon: String)] = [:]
    for entry in list.prefix(40) {
        let dt = Date(timeIntervalSince1970: entry["dt"] as? TimeInterval ?? 0)
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        let dateKey = dateFormatter.string(from: dt)
        let main = entry["main"] as? [String: Any] ?? [:]
        let min = main["temp_min"] as? Double ?? 0
        let max = main["temp_max"] as? Double ?? 0
        let weatherArr = entry["weather"] as? [[String: Any]] ?? []
        let icon = weatherArr.first?["icon"] as? String ?? ""
        if let existing = days[dateKey] {
            days[dateKey] = (min: min(existing.min, min), max: max(existing.max, max), icon: existing.icon)
        } else {
            days[dateKey] = (min: min, max: max, icon: icon)
        }
    }
    let tempUnit = units == "metric" ? "°C" : "°F"
    var count = 0
    for (date, vals) in days.sorted(by: { $0.key < $1.key }) {
        if count >= 5 { break }
        let icon = iconForCode(vals.icon)
        print("\(date): \(icon) \(Int(vals.max))° / \(Int(vals.min))°")
        count += 1
    }
}

func main() {
    if apiKey == "YOUR_API_KEY_HERE" {
        fputs("Please set OPENWEATHER_API_KEY environment variable.\n", stderr)
        exit(1)
    }
    let city = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "London"
    if let weather = getWeather(city: city) {
        displayWeather(weather)
        if let forecast = getForecast(city: city) {
            displayForecast(forecast)
        }
    } else {
        print("Failed to get weather data.")
    }
}

main()
